agent/cache/cache.go

```diff
@@ -613,9 +613,97 @@ func (c *Cache) fetch(key string, r getOptions, allowNew bool, attempt uint, ign
 	metrics.SetGauge([]string{"consul", "cache", "entries_count"}, float32(len(c.entries)))
 	metrics.SetGauge([]string{"cache", "entries_count"}, float32(len(c.entries)))
 
-	tEntry := r.TypeEntry
 	// The actual Fetch must be performed in a goroutine.
-	go func() {
+	go c.launchBackgroundFetcher(entry.GoroutineID, key, r)
+
+	return entry.Waiter
+}
+
+func (c *Cache) launchBackgroundFetcher(goroutineID uint64, key string, r getOptions) {
+	defer func() {
+		c.entriesLock.Lock()
+		defer c.entriesLock.Unlock()
+		entry, ok := c.entries[key]
+		if ok && entry.GoroutineID == goroutineID {
+			entry.GoroutineID = 0
+			entry.Fetching = false
+			c.entries[key] = entry
+		}
+	}()
+
+	var attempt uint
+	for {
+		shouldStop, shouldBackoff := c.runBackgroundFetcherOnce(goroutineID, key, r)
+		if shouldStop {
+			return
+		}
+
+		if shouldBackoff {
+			attempt++
+		} else {
+			attempt = 0
+		}
+		// If we're over the attempt minimum, start an exponential backoff.
+		wait := backOffWait(attempt)
+
+		// If we have a timer, wait for it
+		wait += r.TypeEntry.Opts.RefreshTimer
+
+		select {
+		case <-time.After(wait):
+		case <-c.stopCh:
+			return // Check if cache was stopped
+		}
+
+		// Trigger. The "allowNew" field is false because in the time we were
+		// waiting to refresh we may have expired and got evicted. If that
+		// happened, we don't want to create a new entry.
+		r.Info.MustRevalidate = false
+		r.Info.MinIndex = 0
+
+		// We acquire a write lock because we may have to set Fetching to true.
+		c.entriesLock.Lock()
+
+		entry, ok := c.entries[key]
+		if !ok || entry.Fetching || entry.GoroutineID != goroutineID {
+			// If we don't have an existing entry, return immediately.
+			//
+			// Also if we already have an entry and it is actively fetching, then
+			// return immediately.
+			//
+			// If we've somehow lost control of the entry, also return.
+			c.entriesLock.Unlock()
+			return
+		}
+
+		// Set that we're fetching to true, which makes it so that future
+		// identical calls to fetch will return the same waiter rather than
+		// perform multiple fetches.
+		entry.Fetching = true
+		c.entries[key] = entry
+		metrics.SetGauge([]string{"consul", "cache", "entries_count"}, float32(len(c.entries)))
+		metrics.SetGauge([]string{"cache", "entries_count"}, float32(len(c.entries)))
+		c.entriesLock.Unlock()
+	}
+}
+
+func (c *Cache) runBackgroundFetcherOnce(goroutineID uint64, key string, r getOptions) (shouldStop, shouldBackoff bool) {
+	// Freshly re-read this, rather than relying upon the caller to fetch it
```

Comment:
I also took the opportunity to clean this up and just fetch the data every time, instead of fetching it before calling this function, which had some error prone logic.